### PR TITLE
Upgrade dependencies and fix state-tracking bug

### PR DESCRIPTION
**Pull request by aspyrx**
Previously, the context and props handlers would not be called to indicate that a component's name has changed or it got unmounted. This PR fixes this issue, and adds the appropriate tests.

It also upgrades some outdated dependencies (`istanbul` -> `nyc`, `godaddy-style` -> `eslint-config-godaddy-react`, `react-addons-test-utils` -> `react-test-renderer`, other various version bumps) and removes the requirement for `jsdom` for testing.

If this gets merged, it would probably be a major version bump - it introduces extra calls to lifecycle/change handlers, as well as a new validity state. The `undefined` state existed implicitly before (when a `Validates` did not have a `validates` prop), but it was undocumented. It now has a better-defined meaning documented in the README.

### CODE REVIEW

File: validate.js [hunk 1]
```diff
@@ -1,87 +1,116 @@
...
     this.state = {
-      validates: props.validates,
+      validates: undef,
       valids: {}
     };
   }
 
+  get validates() {
+    // Prefer props over state.
+    const { validates = this.state.validates } = this.props;
```
Comment (indexzero) [inline validate.js]: This is a great ES6-based React pattern I would not have thought of. 💯

File: validates.js [hunk 2]
```diff
@@ -1,67 +1,93 @@
...
 export default class Validates extends React.Component {
+  get propsHandler() {
+    const { onValidChange = noop } = this.props;
+    return onValidChange;
+  }
+
+  get ctxHandler() {
+    const { onValidChange = noop } = this.context;
+    return onValidChange;
+  }
```
Comment (indexzero) [inline validates.js]: You only use these getters once in `onValidChange`. Remove them and simply define them in that function scope.

File: validates.js [hunk 3]
```diff
@@ -1,67 +1,93 @@
...
-  onValidChange (isValid, wasValid) {
-    if (isValid === wasValid) {
-      return;
-    }
-
-    const { onValidChange: propsHandler, name } = this.props;
-    const { onValidChange: ctxHandler } = this.context;
+  onValidChange(isValid, wasValid, oldName) {
+    const { propsHandler, ctxHandler } = this;
+    const { name } = this.props;
```
Comment (indexzero) [inline validates.js]: As per above comment this would be:

``` js
const { name, propsHandler = noop } = this.props;
const { ctxHandler = noop } = this.context;
```

File: validates.js [hunk 4]
```diff
@@ -1,67 +1,93 @@
...
+   * React lifecycle handler called when a component finished updating.
    *
-   * @param {Object} nextProps Component's new props.
+   * @param {Object} prevProps Component's old props.
    */
-  componentWillUpdate (nextProps) {
-    const { validates: wasValid } = this.props;
-    const { validates: isValid } = nextProps;
+  componentDidUpdate(prevProps) {
+    const { validates: wasValid, name: oldName } = prevProps;
```
Comment (indexzero) [inline validates.js]: Nitpick: `prevName` since it came from `prevProps`

File: validate.js [hunk 5]
```diff
@@ -1,87 +1,116 @@
...
-  componentWillUpdate (nextProps, nextState) {
-    const { validates: wasValid = this.state.validates } = this.props;
-    const { validates: isValid = nextState.validates } = nextProps;
+  componentDidUpdate(prevProps, prevState) {
+    const isValid = this.validates;
 
-    this.onValidChange(isValid, wasValid);
+    // Prefer props over state.
+    const { validates: wasValid = prevState.validates, name: oldName } = prevProps;
+    this.onValidChange(isValid, wasValid, oldName);
```
Comment (indexzero) [inline validate.js]: Nitpick: `prevName` since it came from `prevProps`

File: validates.js [hunk 6]
```diff
@@ -1,67 +1,93 @@
...
+   * context handlers will be called first with `undefined` to indicate th old name no longer has validation.
    */
-  onValidChange (isValid, wasValid) {
-    if (isValid === wasValid) {
-      return;
-    }
-
-    const { onValidChange: propsHandler, name } = this.props;
-    const { onValidChange: ctxHandler } = this.context;
+  onValidChange(isValid, wasValid, oldName) {
```
Comment (indexzero) [inline validates.js]: Nitpick: rename `oldName` to `prevName` for consistency with other rename nitpicks. Nitpicks on nitpicks?

**Review by indexzero — CHANGES_REQUESTED** [review]
Would like to see the getters removed before merging. Overall looks like a solid refactor. Definitely a lot of complexity in all the lifecycle edge-cases. Glad to see all the unit tests appear to cover them 👍

File: package.json [hunk 7]
```diff
@@ -18,11 +18,16 @@
...
+    "require": ["babel-register"],
+    "reporter": ["text", "html"]
+  },
   "scripts": {
-    "test": "istanbul cover --report cobertura _mocha -- --compilers js:babel-register",
-    "pretest": "godaddy-js-style-lint *.js ./test",
-    "prepublish": "npm run build",
-    "build": "babel *.js -d lib"
+    "build": "babel *.js -d lib",
+    "lint": "eslint --fix *.js test",
```
Comment (3rd-Eden) [inline package.json:27]: Why are you not using the `eslint-godaddy-react` binary that is installed by the `eslint-config-godaddy-react` package here? This way you don't need to have a the .eslint file in the root of the repository.

File: package.json [hunk 8]
```diff
@@ -47,24 +52,26 @@
...
+    "babel-cli": "^6.24.1",
+    "babel-register": "^6.24.1",
+    "eslint": "^3.19.0",
+    "eslint-config-godaddy-react": "^1.1.1",
+    "eslint-plugin-json": "^1.2.0",
+    "eslint-plugin-mocha": "^4.11.0",
+    "eslint-plugin-react": "^6.10.3",
+    "mocha": "^3.4.2",
+    "nyc": "^11.0.3",
+    "prop-types": "15.x.x",
```
Comment (3rd-Eden) [inline package.json]: could we semver range this in exactly the same way the rest of the dependencies?

File: validate.js [hunk 9]
```diff
@@ -1,87 +1,116 @@
...
     super(props);
 
     this.state = {
-      validates: props.validates,
+      validates: undef,
       valids: {}
     };
   }
 
+  get validates() {
```
Comment (3rd-Eden) [inline validate.js]: JSDoc

File: validate.js [hunk 10]
```diff
@@ -1,87 +1,116 @@
-import React from 'react';
+import { func } from 'prop-types';
 import Validates from './validates';
 
+const undef = void 0;
+
 export default class Validate extends Validates {
```
Comment (3rd-Eden) [inline validate.js]: JSDoc

**Review by 3rd-Eden — DISMISSED** [review]

File: package.json (hunk 7 above)
Comment (aspyrx) [reply package.json:27]: Personally, I find it nice to specify the `.eslint` config at the project root just to make it clear what config this project is using for tools that run `eslint` separately (for example, when Neovim runs it using `neomake`). Of course that's a bit of a selfish reason so I'm happy to remove it.

Comment (indexzero) [reply package.json:27]: The other argument @3rd-Eden is that the `eslint-godaddy-react` binaries have 💩 💩 💩 support for Windows due to how `which` works cross platform. Keeping it this way make contribution more accessible to Windows folks.

**Review by indexzero — APPROVED** [review]
👍 agree with comments from @3rd-Eden, but they could be done post merge.
